File: web/src/model/victron_controller/dashboard/WeatherSocInputs.ts
// @ts-nocheck
import {BaboonGeneratedLatest, BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'
import {WeatherSocDay, WeatherSocDay_UEBACodec} from './WeatherSocDay'
import {WeatherSocTemperatureSource, WeatherSocTemperatureSource_UEBACodec} from './WeatherSocTemperatureSource'

export class WeatherSocInputs implements BaboonGeneratedLatest {
    private readonly _temperature_c: number;
    private readonly _temperature_source: WeatherSocTemperatureSource;
    private readonly _energy_kwh: number;
    private readonly _day: WeatherSocDay;

    constructor(temperature_c: number, temperature_source: WeatherSocTemperatureSource, energy_kwh: number, day: WeatherSocDay) {
        this._temperature_c = temperature_c
        this._temperature_source = temperature_source
        this._energy_kwh = energy_kwh
        this._day = day
    }

    public get temperature_c(): number {
        return this._temperature_c;
    }
    public get temperature_source(): WeatherSocTemperatureSource {
        return this._temperature_source;
    }
    public get energy_kwh(): number {
        return this._energy_kwh;
    }
    public get day(): WeatherSocDay {
        return this._day;
    }

    public toJSON(): Record<string, unknown> {
        return {
            temperature_c: this._temperature_c,
            temperature_source: this._temperature_source,
            energy_kwh: this._energy_kwh,
            day: this._day
        };
    }

    public with(overrides: {temperature_c?: number; temperature_source?: WeatherSocTemperatureSource; energy_kwh?: number; day?: WeatherSocDay}): WeatherSocInputs {
        return new WeatherSocInputs(
            'temperature_c' in overrides ? overrides.temperature_c! : this._temperature_c,
            'temperature_source' in overrides ? overrides.temperature_source! : this._temperature_source,
            'energy_kwh' in overrides ? overrides.energy_kwh! : this._energy_kwh,
            'day' in overrides ? overrides.day! : this._day
        );
    }

    public static fromPlain(obj: {temperature_c: number; temperature_source: WeatherSocTemperatureSource; energy_kwh: number; day: WeatherSocDay}): WeatherSocInputs {
        return new WeatherSocInputs(
            obj.temperature_c,
            obj.temperature_source,
            obj.energy_kwh,
            obj.day
        );
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return WeatherSocInputs.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return WeatherSocInputs.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#WeatherSocInputs'
    public baboonTypeIdentifier() {
        return WeatherSocInputs.BaboonTypeIdentifier
    }
    public static readonly BaboonSameInVersions = ["0.3.0"]
    public baboonSameInVersions() {
        return WeatherSocInputs.BaboonSameInVersions
    }
    public static binCodec(): WeatherSocInputs_UEBACodec {
        return WeatherSocInputs_UEBACodec.instance
    }
}

export class WeatherSocInputs_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: WeatherSocInputs, writer: BaboonBinWriter): unknown {
        if (this !== WeatherSocInputs_UEBACodec.lazyInstance.value) {
          return WeatherSocInputs_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        if (ctx === BaboonCodecContext.Indexed) {
            BinTools.writeByte(writer, 0x01);
            const buffer = new BaboonBinWriter();
            BinTools.writeF64(buffer, value.temperature_c);
            WeatherSocTemperatureSource_UEBACodec.instance.encode(ctx, value.temperature_source, buffer);
            BinTools.writeF64(buffer, value.energy_kwh);
            WeatherSocDay_UEBACodec.instance.encode(ctx, value.day, buffer);
            writer.writeAll(buffer.toBytes());
        } else {
            BinTools.writeByte(writer, 0x00)
            BinTools.writeF64(writer, value.temperature_c);
            WeatherSocTemperatureSource_UEBACodec.instance.encode(ctx, value.temperature_source, writer);
            BinTools.writeF64(writer, value.energy_kwh);
            WeatherSocDay_UEBACodec.instance.encode(ctx, value.day, writer);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): WeatherSocInputs {
        if (this !== WeatherSocInputs_UEBACodec .lazyInstance.value) {
            return WeatherSocInputs_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const header = BinTools.readByte(reader);
        const useIndices = header === 0x01;
        if (useIndices) {
            for (let i = 0; i < 0; i++) {
                BinTools.readI32(reader);
                BinTools.readI32(reader);
            }
        }
        const temperature_c = BinTools.readF64(reader);
        const temperature_source = WeatherSocTemperatureSource_UEBACodec.instance.decode(ctx, reader);
        const energy_kwh = BinTools.readF64(reader);
        const day = WeatherSocDay_UEBACodec.instance.decode(ctx, reader);
        return new WeatherSocInputs(
            temperature_c,
            temperature_source,
            energy_kwh,
            day,
        );
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return WeatherSocInputs_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return WeatherSocInputs_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#WeatherSocInputs'
    public baboonTypeIdentifier() {
        return WeatherSocInputs_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new WeatherSocInputs_UEBACodec())
    public static get instance(): WeatherSocInputs_UEBACodec {
        return WeatherSocInputs_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/weather_soc_inputs.rs
use crate::victron_controller::dashboard::weather_soc_day::WeatherSocDay;
use crate::victron_controller::dashboard::weather_soc_temperature_source::WeatherSocTemperatureSource;

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct WeatherSocInputs {
    pub temperature_c: f64,
    pub temperature_source: WeatherSocTemperatureSource,
    pub energy_kwh: f64,
    pub day: WeatherSocDay,
}

impl PartialEq for WeatherSocInputs {
    fn eq(&self, other: &Self) -> bool {
        self.cmp(other) == std::cmp::Ordering::Equal
    }
}

impl Eq for WeatherSocInputs {}

impl PartialOrd for WeatherSocInputs {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for WeatherSocInputs {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        match self.temperature_c.total_cmp(&other.temperature_c) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.temperature_source.cmp(&other.temperature_source) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.energy_kwh.total_cmp(&other.energy_kwh) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.day.cmp(&other.day) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        std::cmp::Ordering::Equal
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for WeatherSocInputs {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for WeatherSocInputs {
    fn encode_ueba(&self, ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let value = self;
        if ctx.use_indices() {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x01)?;
            let mut buffer: Vec<u8> = Vec::new();
            value.temperature_c.encode_ueba(ctx, &mut buffer)?;
            value.temperature_source.encode_ueba(ctx, &mut buffer)?;
            value.energy_kwh.encode_ueba(ctx, &mut buffer)?;
            value.day.encode_ueba(ctx, &mut buffer)?;
            writer.write_all(&buffer)?;
        } else {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x00)?;
            value.temperature_c.encode_ueba(ctx, writer)?;
            value.temperature_source.encode_ueba(ctx, writer)?;
            value.energy_kwh.encode_ueba(ctx, writer)?;
            value.day.encode_ueba(ctx, writer)?;
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for WeatherSocInputs {
    fn decode_ueba(ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let (_header, index) = <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::read_index(ctx, reader)?;
        if ctx.use_indices() {
            assert_eq!(index.len(), <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::index_elements_count(ctx) as usize);
        }
        let temperature_c = crate::baboon_runtime::bin_tools::read_f64(reader)?;
        let temperature_source = WeatherSocTemperatureSource::decode_ueba(ctx, reader)?;
        let energy_kwh = crate::baboon_runtime::bin_tools::read_f64(reader)?;
        let day = WeatherSocDay::decode_ueba(ctx, reader)?;
        Ok(WeatherSocInputs {
            temperature_c,
            temperature_source,
            energy_kwh,
            day,
        })
    }
}